scripts/perf/scheduler_dag_execution_timing.py

```diff
@@ -0,0 +1,221 @@
+
+    # from airflow.executors.local_executor import LocalExecutor
+
+    # Change this to try other executors
+    Executor = MockExecutor
+
+    class ShortCircutExecutor(ShortCircutExecutorMixin, Executor):
+        pass
+
+    return ShortCircutExecutor
+
+
+def reset_dag(dag, num_runs, session):
+    import airflow.models
+    from airflow.utils import timezone
+    from airflow.utils.state import State
+
+    DR = airflow.models.DagRun
+    DM = airflow.models.DagModel
+    TI = airflow.models.TaskInstance
+    TF = airflow.models.TaskFail
+    dag_id = dag.dag_id
+
+    session.query(DM).filter(DM.dag_id == dag_id).update({'is_paused': False})
+    session.query(DR).filter(DR.dag_id == dag_id).delete()
+    session.query(TI).filter(TI.dag_id == dag_id).delete()
+    session.query(TF).filter(TF.dag_id == dag_id).delete()
+
+    next_run_date = dag.normalize_schedule(dag.start_date or min(t.start_date for t in dag.tasks))
+
+    for _ in range(num_runs):
+        next_run = dag.create_dagrun(
+            run_id=DR.ID_PREFIX + next_run_date.isoformat(),
+            execution_date=next_run_date,
+            start_date=timezone.utcnow(),
+            state=State.RUNNING,
+            external_trigger=False,
+            session=session,
+        )
+        next_run_date = dag.following_schedule(next_run_date)
+    return next_run
+
+
+def pause_all_dags(session):
+    from airflow.models.dag import DagModel
+    session.query(DagModel).update({'is_paused': True})
+
+
+@click.command()
+@click.option('--num-runs', default=1, help='number of DagRun, to run for each DAG')
+@click.option('--repeat', default=3, help='number of times to run test, to reduce variance')
+@click.argument('dag_ids', required=True, nargs=-1)
+def main(num_runs, repeat, dag_ids):
+    """
+    This script will run the SchedulerJob for the specified dags "to completion".
+
+    That is it creates a fixed number of DAG runs for the specified DAGs (from
+    the configured dag path/example dags etc), disable the scheduler from
+    creating more, and then monitor them for completion. When the file task of
+    the final dag run is completed the scheduler will be terminated.
+
+    The aim of this script is to have a benchmark for real-world scheduler
+    performance -- i.e. total time take to run N dag runs to completion.
+
+    Care should be taken that other limits (DAG concurrency, pool size etc) are
+    not the bottleneck. This script doesn't help you in that regard.
+
+    It is recommended to repeat the test at least 3 times (`--repeat=3`) so
+    that you can get somewhat-accurate variance on the reported timing numbers,
+    but this can be disabled for longer runs if needed.
+    """
+    # Disallow the scheduler to create new dag_runs - we've already created all
+    # the ones we want it to process. Easier than editing end date
+    os.environ['AIRFLOW__SCHEDULER__USE_JOB_SCHEDULER'] = 'False'
```

Comment:
This changes the number of queries and falsifies results. When my optimizations are introduced, DagFileProcessor will always call 5 queries regardless of the number of Dags in the file.

However, if this option is disabled, the number of queries will vary over time. This has a big impact on performance.

Reply:
Does it? Link to where in the code this is still used.

The main reason this is here is to remove sleeps in 1.10 releases

Reply:
https://github.com/apache/airflow/blob/master/airflow/jobs/scheduler_job.py#L514

PS. I often tested DAGs that had schedule_interval = None, to reduce noise in logs.

Reply:
https://github.com/apache/airflow/pull/7502/files where in there will this change the number of queries? After your optimization lands we should always call `DagFileProcessor.create_dag_run` with dags_runs prepopulated.

In short: same number of queries either way.

Reply:
https://github.com/apache/airflow/blob/master/airflow/jobs/scheduler_job.py#L540-L550
https://github.com/apache/airflow/blob/master/airflow/jobs/scheduler_job.py#L589
I would like to optimize it later because we have a lot of PR for the same class.

Reply:
So it's one less query per time processing the DAG. So it will make this benchmark slightly faster overall.

However this behaviour (if not this implementation) is a key part of this script -- to get a _repeatable_ number of "how fast will the scheduler process X dag runs to completion".

Do you have any ideas how I could achieve this behaviour (to only allow a specific number of dag runs) without needing to edit the dag file?

I could monkey-patch `schedule_interval` on the DAG class to be a dynamic property that counts how many runs are created, and after N is reached returns none instead.

Reply:
Monkey patching may not work on OSX with Py 3.8 -- they changed the default to spawn a new process didn't they :(

Reply:
@mik-laj Please take a look at the new approach.

Reply:
I think these tests will never be reliable unless we test it in isolation. We have two loops: SchedlerJob and DagFileProcessorManager.  We should write performance tests for the two loops separately.

Reply:
Och. We also have executor loop. It 's synchronized with SchedulerLoop, but we can also run this loop in isolation.

Reply:
What is wrong with modifying the DAG file?  I think it is worth preparing a ready set of files that will test different situations and will be used to compare performance. We won't be able to detect regression if we're testing only a different cases.

Reply:
Then in DagFile we can specify start_date and end_date and schedule_interval. This will provide us with a constant number of DagRun. This is launching us with a fixed DagRun number.

Reply:
> I think these tests will never be reliable unless we test it in isolation.

Yes, there is value in testing the parts in isolation, but the purpose of this test is to have a way of measuring "Total Scheduler Overhead" -- how quickly can the scheduler and all it's parts get out of the way and run the actual tasks. Without measuring the whole scheduler performance we could make a part 1000x faster, but if it's not where the scheduler is waiting, it won't make tasks start any quicker.

The reason to not want to edit the dag file is so that this can be run against _any_ dag, not just a specially prepared one.

Reply:
For instance if we want to say "from 1.10 to 2.0 your tasks will now be scheduled 3x faster" we need this sort of benchmark.

Reply:
But this monkey patch approach doesn't work -- the `DAG` object isn't persistent, so each process sees it has created 0 dag runs.

Okay, setting end_date it is. The rest of the script still works even if you don't set the end date, it will just create a few more dag_runs (and may start scheduling those tasks) is all.